### DISCUSSION TITLE
Using an Openbox-style "client-list-menu" to compensate for a lack of a working workspaces widget in waybar and the like

### DISCUSSION THREAD

**Discussion by jjramsey** [labwc/labwc#1945] [Ideas]
It's been mentioned before that there's a lack of a working workspace widget for Labwc in apps like waybar because the Wayland protocols to support it are still in flux (https://github.com/labwc/labwc/issues/881). Doing like some other compositors have done by using some custom protocol is out of scope for Labwc (https://github.com/labwc/labwc/issues/293).

Openbox's "client-list-menu" does a similar job to a workspace widget, in that it lets one know which windows are on which workspaces, and one can, say, put a checkmark next to the current workspace as an indicator and still be within the Openbox menu spec. Furthermore, implementing another part of Openbox's menu spec should be in-scope for Labwc. This would basically be a workaround for the lack of a workspace widget that would fit within Labwc's remit and not involve issues with custom protocols.

There was an earlier discussion that broached implementing a client list feature (https://github.com/labwc/labwc/discussions/986), but it didn't discuss its use as a possible substitute for a workspace widget.

**Comment by johanmalm** [comment]
Yes, I do like the idea of this. We ought to do it.

**Comment by droc12345** [comment]
Doesn't the window switcher already do that, ie show all apps on all workspaces, if using custom layout.
It just doesn't group them by workspace, which is an enhancement that I've considered. 

Or am I missing something, I never used client-list-menu when I was using openbox.

**Reply by stefonarch** [reply]
The switcher can [sort](https://labwc.github.io/labwc-config.5.html#entry_windowswitcher_show=_preview=_outlines=_allworkspaces=) them out by workspace already, this info should only be accessible by panels and other switchers.

**Reply by jjramsey** [reply]
> The switcher can [sort](https://labwc.github.io/labwc-config.5.html#entry_windowswitcher_show=_preview=_outlines=_allworkspaces=) them out by workspace already.

Doesn't look like the manpage to which you linked indicates any *sorting* by workspace, just that the workspace is a field that can be shown.

**Comment by droc12345** [comment]
Actually after I made my comment I started thinking, it would be nice to have something like client-list-menu
for a couple of reasons, 1. it can sort by workspace, 2. It allows using the mouse.
While window switcher will show workspace (and monitor if configured) you can't use the mouse with it.

As I say I never used client-list-menu before does it allow selection of windows or is it show only, no selection?
If it's selectable and one can use the mouse, then it would be a nice addition, especially if the existing code can be reused.

**Comment by droc12345** [comment]
I've been looking at the few references to client-list-{combined}-menu and I do like the idea.
We would probably want to add an x/y placement option along with the current atCursor yes/no (openbox has a placement option)
if position is set it would override atCursor except for root-menu and client-menu.
This would allow placement wherever the end user wanted, including center.

**Reply by johanmalm** [reply]
Yes, I was thinking the same ref x/y.

I have made a start on implementing menu titles. Think we need that first.

**Comment by droc12345** [comment]
Johan
I have parts of it done, I took your menu patches, added what was needed to make non-parent pipemenus work
and added x,y placement options (if enabled overrides all menus except for client-menu which follows atCursor)

Initially I wasn't really sure how to handle client-list-{combined}-menu so I went looking at the openbox docs and found
that there are several _internal_ menus:

1. client-list-combined-menu 
2. client-list-menu 
3. client-menu 
4. client-send-to-menu 
5. client-layer-menu

We provide a root-menu and client-menu (unless overridden by menu.xml).
Then the question is, do we do all of them? 
edit to add: 3,4,5 are all view centric, will only work on current active window.

The advantages is the compositor knows all the details of all processes, windows, etc and it's easy to have 
send-to-menu show all workspaces to send the application to, for example.
Layer allows setting the layer per window, top, bottom, normal.

client-list-menu shows all applications, with each workspace being a submenu
client-list-combined-menu shows all apps on same menu, using separator labels for each workspace.

**Reply by johanmalm** [reply]
droc
Thanks for working on this.
I'm "out of action" for another week.
Will take a look at your work then if that is okay.

**Reply by droc12345** [reply]
Not a problem, I'm still wrangling with a builtin dynamic menu.

**Comment by Consolatis** [comment]
Note:
- #2030

Along the already posted PR for sfwbar there is also a working implementation for waybar ~~on my disk, need to clean it up and send a PR~~.
Testing would be appreciated.

I do think that labwc generated dynamic menus are still useful though, not everybody uses a panel and some people might prefer dynamic menus. The cosmic-workspaces protocol also only allows selecting the active workspace, it does not allow assigning windows to workspaces.

Edit:
Waybar PR opened, link in the labwc PR linked above.

**Comment by droc12345** [comment]
Ok, got it working, current view of what's what
M - maximized
m - minimized
F - fullscreen 

Blue items are workspaces and it's currently on the web workspace

"Go There" takes you to the workspace
clicking on app title takes you to that view.

And keep in mind, that this is still a WIP.

![test](/../../../droc12345/images/blob/main/data/client-list.jpg?raw=true)

Right now I have the code split into 4 commits 
1. johanmalms #1976 draft
2. pipemenu changes to allow non-parent
3. add x,y menu placement (based on openbox and like openbox set globally)
4. client-list-combined-menu (based on openbox)

added another commit, client-send-to-menu, handles all workspaces (client-list workspace submenu)

![test](/../../../droc12345/images/blob/main/data/sendto.jpg?raw=true)

ETA: added 2 more commmits
5. client-send-to-menu - show all workspaces as well as left/right
6. separator label changes - text color, label font, justification (center)

client list, w title centered, using text font and oblique slanted font
![test](/../../../droc12345/images/blob/main/data/client-list-2.jpg?raw=true)

**Comment by droc12345** [comment]
I have started to carry patches on my own, they work against the latest labwc git

**Comment by droc12345** [comment]
Client list combined menu has been merged.

**Comment by fungilife** [comment]
sfwbar now shows workpace widget but buttons out of order and a bit larger than expected.  Possibly a sfwbar config issue not yet discovered.  Otherwise it seems functional.   
Noticed on rc96 c413e65a...
Currently on rc101 a5d89a2e...
on joborun labwc-git

**Reply by johanmalm** [reply]
Yes.

I'm currently using 

```

  pager {
    preview = false
    sort = true
    rows = 1
  }
```

Seems to work fine.

**Reply by fungilife** [reply]
Thank you for the tip, I wonder whether having ws labeled with characters instead of numbers it makes the difference,  Whether I turn sort on off use the pins definition there is no way to get them in the right order.  When using keybinds/mouse to switch around both labels and order seem fine, when the pager picks up the labels it orders them differently.
